source/common/upstream/maglev_lb.cc

```diff
@@ -3,38 +3,60 @@
 namespace Envoy {
 namespace Upstream {
 
-MaglevTable::MaglevTable(const HostVector& hosts, uint64_t table_size) : table_size_(table_size) {
+MaglevTable::MaglevTable(const HostsPerLocality& hosts_per_locality,
+                         const LocalityWeightsConstSharedPtr& locality_weights, uint64_t table_size)
+    : table_size_(table_size) {
   // TODO(mattklein123): The Maglev table must have a size that is a prime number for the algorithm
   // to work. Currently, the table size is not user configurable. In the future, if the table size
   // is made user configurable, we will need proper error checking that the user cannot configure a
   // size that is not prime (the result is going to be an infinite loop with some inputs which is
   // not good!).
   ASSERT(Primes::isPrime(table_size));
-  if (hosts.empty()) {
-    return;
-  }
+
+  // Compute host weight combined with locality weight where applicable.
+  const auto effective_weight = [&locality_weights](uint32_t host_weight,
+                                                    uint32_t locality_index) -> uint32_t {
+    if (locality_weights == nullptr || locality_weights->empty()) {
+      return host_weight;
+    } else {
+      return host_weight * (*locality_weights)[locality_index];
+    }
+  };
 
   // Compute maximum host weight. If this is zero, we are doing unweighted Maglev.
   uint32_t max_host_weight = 0;
-  for (const auto& host : hosts) {
-    max_host_weight = std::max(host->weight(), max_host_weight);
+  uint32_t total_hosts = 0;
+  for (uint32_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      max_host_weight =
+          effective_weight(std::max(effective_weight(host->weight(), i), max_host_weight), i);
```

Comment:
Having a hard time following the math here. Can you add a comment?

Reply:
That's because there's some typo there, will fix.